### PR TITLE
feat: add check unused code command

### PR DESCRIPTION
**Pull request by incendial**
### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix
[ ] New rule
[ ] Changes an existing rule
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[X] Other, please explain: new command

### What changes did you make? (Give an overview)

This implementation is a first iteration of unused code check, it checks for top level declarations usage, such as:
- top level variables
- top level functions
- enums
- classes
- mixins
- extensions
- type aliases (a.k.a. typedefs)

In the next iterations we should add checks for specific class method / properties / getters usage, but this command already brings a lot of value to our users, so let's not block the release on this.

### CODE REVIEW

**Comment by incendial** [comment]
@diegoveloper @lukepighetti hope it's the command you were looking for. Do you want to "beta-test" it?

**Comment by diegoveloper** [comment]
> @diegoveloper @lukepighetti hope it's the command you were looking for. Do you want to "beta-test" it?

Ohh nice!! We'll try the next week

**Review by vlkonoshenko — APPROVED** [review]

File: lib/src/analyzers/unused_code_analyzer/models/unused_code_issue.dart [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import 'package:meta/meta.dart';
+import 'package:source_span/source_span.dart';
+
+/// Represents an issue detected by the unused code check.
+@immutable
```
Comment (dkrutskikh) [inline lib/src/analyzers/unused_code_analyzer/models/unused_code_issue.dart:5]: @incendial please remove

File: lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart [hunk 2]
```diff
@@ -0,0 +1,53 @@
+// ignore_for_file: public_member_api_docs
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+
+class PublicCodeVisitor extends GeneralizingAstVisitor<void> {
+  final Set<Element> topLevelElements = {};
+
+  PublicCodeVisitor();
```
Comment (dkrutskikh) [inline lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart:10]: @incendial why you define empty constructor?

File: lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  @override
+  void visitCompilationUnitMember(CompilationUnitMember node) {
+    if (node is FunctionDeclaration) {
+      if (node.name.name == 'main' || _hasPragmaAnnotation(node.metadata)) {
```
Comment (dkrutskikh) [inline lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart:15]: @incendial do we use similar code for detect unused files?

File: lib/src/analyzers/unused_code_analyzer/reporters/reporters_list/json/unused_code_json_reporter.dart [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import 'package:meta/meta.dart';
+
+import '../../../../../reporters/models/json_reporter.dart';
+import '../../../models/unused_code_file_report.dart';
+import '../../../models/unused_code_issue.dart';
+
+/// Unused code JSON reporter.
+///
+/// Use it to create reports in JSON format.
+@immutable
```
Comment (dkrutskikh) [inline lib/src/analyzers/unused_code_analyzer/reporters/reporters_list/json/unused_code_json_reporter.dart:13]: @incendial please remove

File: website/docs/cli/check-unused-code.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
+# Check unused code
```
Comment (dkrutskikh) [inline website/docs/cli/check-unused-code.md:1]: @incendial please update overview.md

File: lib/src/cli/commands/check_unused_code_command.dart [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      );
+  }
+
+  void _usesExitOption() {
+    argParser
+      ..addSeparator('')
+      ..addFlag(
+        FlagNames.fatalOnUnused,
+        help: 'Treat find unused file as fatal.',
+// TODO(dkrutrkikh): activate on next major version
```
Comment (dkrutskikh) [inline lib/src/cli/commands/check_unused_code_command.dart:84]: @incendial I think for new commands we can activate immediately

**Review by dkrutskikh — CHANGES_REQUESTED** [review]

**Comment by dkrutskikh** [comment]
@incendial please add test like test/src/cli/commands/check_unused_files_command_test.dart and update readme.md at line 123

File: website/docs/cli/check-unused-code.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+$ dart run dart_code_metrics:metrics check-unused-code lib
+
+# or for a Flutter package
+$ flutter pub run dart_code_metrics:metrics check-unused-code lib
+```
+
+Full command description:
+
+```text
+Usage: metrics check-unused-l10n [arguments] <directories>
```
[the remark this answers is not in the document]
Comment (incendial) [reply website/docs/cli/check-unused-code.md]: Fixed

File: lib/src/analyzers/unused_code_analyzer/models/unused_code_issue.dart (hunk 1 above)
Comment (incendial) [reply lib/src/analyzers/unused_code_analyzer/models/unused_code_issue.dart]: Fixed

File: lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart (hunk 2 above)
Comment (incendial) [reply lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart]: Removed

File: lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart (hunk 3 above)
Comment (incendial) [reply lib/src/analyzers/unused_code_analyzer/public_code_visitor.dart]: Moved to utils

File: lib/src/analyzers/unused_code_analyzer/reporters/reporters_list/json/unused_code_json_reporter.dart (hunk 4 above)
Comment (incendial) [reply lib/src/analyzers/unused_code_analyzer/reporters/reporters_list/json/unused_code_json_reporter.dart]: Removed

File: website/docs/cli/check-unused-code.md (hunk 5 above)
Comment (incendial) [reply website/docs/cli/check-unused-code.md:1]: Done

File: lib/src/cli/commands/check_unused_code_command.dart (hunk 6 above)
Comment (incendial) [reply lib/src/cli/commands/check_unused_code_command.dart:84]: Let's keep all commands in sync

File: lib/src/analyzers/unused_code_analyzer/unused_code_config.dart [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import 'package:meta/meta.dart';
+
+import '../../config_builder/models/analysis_options.dart';
+
+/// Represents raw unused code config which can be merged with other raw configs.
+@immutable
```
[the remark this answers is not in the document]
Comment (incendial) [reply lib/src/analyzers/unused_code_analyzer/unused_code_config.dart]: Removed

**Review by dkrutskikh — APPROVED** [review]

**Comment by diegoveloper** [comment]
hey @incendial , thanks for this new command, one question, is the `excluding files` working?

I ran the command : `flutter pub run dart_code_metrics:metrics check-unused-code lib`   (4.10.0-dev.1).

And the output contains a lot of `*.g.dart:` files. Checking the documentation   https://dartcodemetrics.dev/docs/cli/check-unused-code
 
Screen Shot 2022-01-18 at 11 49 14 AM

It says that the it's excluding those files by default.

**Comment by dkrutskikh** [comment]
@diegoveloper can we move discussion to the issue https://github.com/dart-code-checker/dart-code-metrics/issues/653 ?

**Comment by incendial** [comment]
@diegoveloper hi, this should fix the problem https://github.com/dart-code-checker/dart-code-metrics/pull/654

**Pull request merged**